app/core/components/optimizer/greedymax.py

```diff
@@ -1,10 +1,139 @@
 # Copyright (c) 2016-2022 Association of Universities for Research in Astronomy, Inc. (AURA)
 # For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+from typing import Mapping
+
+from lucupy.minimodel.program import ProgramID
+
+from app.core.calculations import GroupData, ProgramInfo
+from app.core.plans import Plan, Plans
 from .base import BaseOptimizer
 
+import numpy as np
+import astropy.units as u
 
-class GreedyMax(BaseOptimizer):
+
+class GreedyMaxOptimizer(BaseOptimizer):
     """
     GreedyMax is an optimizer that schedules the visits in a greedy fashion.
     """
-    ...
+
+    def __init__(self, seed=42):
+        self.groups = []
+
+    @staticmethod
+    def _allocate_time(plan: Plan, obs_time: timedelta) -> datetime:
+        """
+        Allocate time for an observation inside a Plan
+        This should be handled by the optimizer as can vary from algorithm to algorithm
+        """
+        # Get first available slot
+        start = plan.start
+        if len(plan.visits) > 0:
+            start = plan.visits[-1].start_time + plan.visits[-1].time_slots * plan.time_slot_length
+        # for v in plan.visits:
+        #     # delta = v.start_time - start + obs_time
+        #     start = v.start_time + v.time_slots * plan.time_slot_length
+        return start
+
+
+    def setup(self, program_info: Mapping[ProgramID, ProgramInfo]) -> GreedyMaxOptimizer:
+        """
+        Preparation for the optimizer e.g. create chromosomes, etc.
+        """
+        self.groups = []
+        # self.scores = []
+        for p in program_info.values():
+            self.groups.extend([g for g in p.group_data.values() if g.group.is_observation_group()])
+            # Suggestion from Seb
+            # self.groups.extend([g for g in p.group_data.values() if all(x.is_observation_group() for x in g.group.children)])
+            # self.scores.extend([s for s in p.group_data])
+        return self
+
+    def _run(self, plans: Plans):
+
+        print()
```

Comment:
I agree with Sergio that we shouldn't be including `print` statements in PRs that we are going to merge, and should instead be logging.